### DISCUSSION TITLE
Precedence with two competing optional() tokens from grammar rule and external scanner

### DISCUSSION THREAD

**Discussion by mikehaertl** [tree-sitter/tree-sitter#1619] [Q&A] [answered]
I'm working on an [external scanner for phpdoc](https://github.com/claytonrcarter/tree-sitter-phpdoc/pull/10).

The external scanner is used to improve scanning of tokens that are used as `(description (text))` nodes. For the most part it already works quite well but I'm stuck at one point when parsing @var tags.

The expected result here is:

```
/** @var int $foo bar baz */
---
(document
    (tag
        (tag_name)
        (type_list (primitive_type))
        (variable_name (name))
        (description (text))))
```

But I get this instead:
```
(document
    (tag
        (tag_name)
        (type_list (primitive_type))
        (description (text))))
```

I.e. it's missing the `(variable_name (name))` node.

<details>
  <summary>Debug output:</summary>

```
tests:
new_parse
process version:0, version_count:1, state:1, row:1, col:0
lex_internal state:4, row:1, column:0
  consume character:'/'
  consume character:'*'
  consume character:'*'
lexed_lookahead sym:_begin, size:3
shift state:2
process version:0, version_count:1, state:2, row:1, col:3
lex_external state:2, row:1, column:3
  skip character:' '
lex_internal state:7, row:1, column:3
  skip character:' '
  consume character:'@'
  consume character:'v'
  consume character:'a'
  consume character:'r'
lexed_lookahead sym:tag_name, size:5
shift state:78
process version:0, version_count:1, state:78, row:1, col:8
lex_internal state:0, row:1, column:8
  skip character:' '
  consume character:'i'
  consume character:'n'
  consume character:'t'
  consume character:'i'
  consume character:'n'
  consume character:'t'
lexed_lookahead sym:int, size:4
shift state:14
process version:0, version_count:1, state:14, row:1, col:12
lex_external state:2, row:1, column:12
  skip character:' '
  consume character:'$'
  consume character:'f'
  consume character:'o'
  consume character:'o'
  consume character:' '
  consume character:'b'
  consume character:'a'
  consume character:'r'
  consume character:' '
  consume character:'b'
  consume character:'a'
  consume character:'z'
  consume character:' '
  consume character:'*'
lexed_lookahead sym:text, size:14
reduce sym:primitive_type, child_count:1
reduce sym:_types, child_count:1
reduce sym:type_list, child_count:1
shift state:36
process version:0, version_count:1, state:36, row:1, col:26
lex_external state:2, row:1, column:26
  consume character:'*'
lex_internal state:7, row:1, column:26
  consume character:'*'
  consume character:'/'
lexed_lookahead sym:_end, size:2
reduce sym:description, child_count:1
reduce sym:_var_tag, child_count:3
reduce sym:tag, child_count:1
shift state:161
process version:0, version_count:1, state:161, row:1, col:28
lex_internal state:0, row:1, column:28
lexed_lookahead sym:end, size:0
reduce sym:document, child_count:3
accept
done
```

</details>

Here first the most relevant parts from the `grammar.js`:

```js
module.exports = grammar({
  externals: $ => [
    $.text,
    $.text_in_inline_tag,
  ],
  rules: {
    document: $ => seq(
      $._begin,
      optional($.description),
      repeat($.tag),
      $._end,
    ),

    _begin: $ => token(seq('/**', repeat('*'))),

    description: $ => repeat1(
      choice(
        $.text,
        $.inline_tag
      )
    ),
    tag: $ => choice(
      // ...
      $._param_tag,
      $._property_tag,
      $._return_tag,
      $._var_tag,
      // ...
    )
    // ..
    _var_tag: $ => seq(
      alias('@var', $.tag_name),
      $._type,
      optional($.variable_name), 
      optional($.description),
    ),
    // ...
    variable_name: $ => seq('$', $.name),
    name: $ => /[_a-zA-Z\u00A1-\u00ff][_a-zA-Z\u00A1-\u00ff\d]*/,

```

Notice the two consecutive `optional()` tokens in `_var_tag`. 

Debug ouput shows, that after parsing `$._type` it  does not even _try_ to parse `$.variable_name` but instead _immediately_ calls my [external scanner](https://github.com/mikehaertl/tree-sitter-phpdoc/blob/1c4e0a7b4eeae607fcdcac816a45d12304605584/src/scanner.c). And for the scanner `$foo bar baz` is a valid description text so it eats up the variable name. I tried to add e.g. `prec()` in different combinations but it had no effect at all (which shows that I probably totally misunderstand what it's meant for):

 * `optional(prec(1, $.variable_name))`
 * `prec(1, optional($.variable_name))`
 * `optional(prec(-1, $.description))`,

So why is scanning of `$.variable_name` skipped and how can I prevent that?

**Comment by ahelwer** [comment, marked as answer]
Your external scanner will always be called immediately if any external token is valid at that position. This is good! It makes it a powerful override of lexing behavior. You can either move the second token into your external scanner or keep some state in your scanner like have_just_seen_token then return false if asked to find that token when your state variable is true and the text starts with $ or something. The same effect could possibly be accomplished without a state variable, as long as $.text is only ever encountered after an optional list of variables.

**Reply by mikehaertl** [reply]
Thanks but I don't quite understand yet. By "second token" do you mean the rule for `$.variable_name`? It's actually the first of the two `optional()` tokens and I wonder why it's skipped. It should match the `$foo` variable and leave the rest for my scanner. I'd prefer if I could leave out scanning for variable names from my scanner as it's much easier to express with regex.

Could you also explain how the `have_just_seen_token` would work? What token has it seen? The scanner is only called once and eats up the full string. There is no second call.

**Reply by ahelwer** [reply]
In general: it's important when using external scanners to develop a "theory of mind" about how tree-sitter works and when it will call into your external scanner. It's good to watch parse debug output to slowly build this understanding, which I think you've been doing, which is good. 

Tree-sitter will always call into your external scanner when you have a variable name in a doc comment because the variable name is optional, and the token after it is an external token. You have a few options as I see it:

- If your $.text external token is only ever used following a possible list of variable names, you have to detect whether it's a variable name in your external scanner and return false if so
- Move $.text parsing into your grammar itself so you can use lexical/parse/dynamic precedence functions
- if $.text is used in multiple places in your grammar, actually forget that state variable idea - just make this $.text into $.comment_text or something so it's only used in one place.

**Reply by mikehaertl** [reply]
Ok, thanks I think I got it.

Unfortunately it means that the situation is more complex than I though. Indeed a description text (`$.text`) can appear all over the place and can even start with a variable name like `$foo`. We had lexical rules before but they are too limited as we also need to parse inline tags that can be embedded in some description texts, like `some text {@see OtherType with its own text}  more text`.

Anyway - I'll have to spend more time on my scanner (just when I thought I was almost done as most tests already pass ;) ).

**Reply by ahelwer** [reply]
You can definitely add an additional external token with the rule that it cannot start with `$foo`, like `$.text_cannot_start_with_var`.

**Reply by mikehaertl** [reply]
But text *can* start with a var like `$foo` . It's just that special situation with `@var` where an optional variable name at the beginning needs to be captured as `variable_name` not `description`. Here all valid combinations for `@var`:

* `@var int some description` (type = int, description = some description)
* `@var int $foo some description` (type = int, varname = $foo, description = some description)
* `@var int $foo $bar some desc` (type = int, varname = $foo, description = $bar some desc)

I will rather try to move `$.variable_name` into the scanner. ~~`regex.h` can surely be used there, too?~~ (Update: nevermind - regex make no sense here due to how the lexer works)

**Reply by ahelwer** [reply]
It makes sense to split up all those cases into different tokens in your external scanner, present at different parts of your grammar. Then you can branch your external scanner behavior based on which token is marked as valid. If you want them all to appear as `$.text`, you can use `alias($.text_variant, $.text)`.

**Reply by mikehaertl** [reply]
> You can definitely add an additional external token with the rule that it cannot start with `$foo`, like `$.text_cannot_start_with_var`.

I finally got why this helps me. I now have these externals:

```
  externals: $ => [
    $.text,
    $._text_cannot_start_with_varname,
  ],
```
Then I defined 2 types of descriptions:
```
    description: $ => repeat1(
      choice(
        $.text,
        $.inline_tag
      )
    ),

    _description_cannot_start_with_varname: $ => alias(
      repeat1(
        choice(
          alias($._text_cannot_start_with_varname, $.text),
          $.inline_tag
        )
      ),
      $.description
    ),
```
which I then can use in the `@var` rule like so:
```
    _var_tag: $ => seq(
      alias('@var', $.tag_name),
      $._type,
      optional(
        choice(
          $._description_cannot_start_with_varname,
          $.variable,
          seq($.variable_name, $.description)
        )
      )
    ),
```

Now it matches all permutations:
```
/** @var int $foo **/
/** @var int $foo description **/
/** @var int $foo $description starting with varname **/
/** @var int description only **/
```

**Reply by ahelwer** [reply]
Happy it worked out!
